manuals/vshadowmount.1
.Dd July 14, 2026
.Dt VSHADOWMOUNT 1
.Os
.Sh NAME
.Nm vshadowmount
.Nd mounts a Volume Service Snapshot (VSS) volume
.Sh SYNOPSIS
.Nm vshadowmount
.Op Fl o Ar offset
.Op Fl X Ar extended_options
.Op Fl hvV
.Ar source
.Ar mount_point
.Sh DESCRIPTION
.Nm vshadowmount
is a utility to mount a Volume Service Snapshot (VSS) volume
.Pp
.Nm vshadowmount
is part of the
.Nm libvshadow
package.
.Nm libvshadow
is a library to access the Volume Shadow Snapshot (VSS) format
.Pp
.Ar source
a Volume Service Snapshot (VSS) volume
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl o Ar offset
specify the volume offset in bytes
.It Fl v
verbose output to stderr, while vshadowmount will remain running in the \
foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# vshadowmount image1.raw /mnt/vshadow
vshadowmount 20260714
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr vshadowinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libvshadow/issues
.Sh COPYRIGHT
Copyright (C) 2011-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* rename libvshadow_volume_get_store to libvshadow_volume_get_store_by_index
* fix issue with multi threaded read test
  - store descriptor initialization?
* add error tolerance for missing NTFS backup volume header
* remove hard coded 0x4000 block size
  - use or pass io handle block size instead
* add tests
  - create test image
  - create test plan
  - test and add ignore overlay when reading from next store
* multi-threading support
  - what about abort value?
* beta
  - fix open issues
* add libfdata and libfcache for caching ?

API:
* allow access to block list
  - finish and expose libvshadow_store_get_block()
  - split block list and run-time btree
* get operating machine string
* get service machine string
* get size
* is open
* finish store close function

Error tollerance:
* add support for catalog entry 3 defined before 2

Format:
* how are multiple copies per set stored?
* the store file can be fragmented
  - block range list defines the fragments ?
* what about trailing volume data that windows is hiding 0x1000 bytes in size ?
  also seen in BDE, so likely something of NTFS

Tools:
* vshadowinfo
  - print attribute flags description
* vshadowmount
  - dokan fix volume size hints

20130503
* see `git log' for more recent change log
* added store read from file IO handle function
* vshadowmount small changes

20130501
* worked on vshadowmount Dokan support

20130417
* bug fix in dependencies

20130416
* code clean up

20130413
* updated dependencies
* updates and bug fixes in pyvshadow

20130412
* updated README files
* updated dependencies
* added tests directory

20130331
* updated dependencies
* code clean up
* pyvshadow updates
* vshadowtools now detect if there is a VSS signature first and bail out with
  a different error if not

20130304
* added PackageMaker files

20130217
* updated include/types.h

20130205
* fixed typo in vhsadowmount

20130131
* worked on pyvshadow

20130130
* worked on exposing block descriptors via vshadowinfo
* worked on pyvshadow

20130129
* worked on exposing block descriptors via API
* worked on pyvshadow

20130128
* worked on exposing block descriptors via API

20130127
* worked on exposing block descriptors via API
* removed LIBVSHADOW_STORE_FLAG_IO_HANDLE_MANAGED flags

20130125
* worked on exposing block descriptors via API

20130113
* 2013 update
* updated dependencies
* updated msvscpp files
* added vshadowmount.1 man page

20121122
* pyvshadow: code clean up

20121118
* pyvshadow: switched to memory_allocate and memory_free to
  PyMem_Malloc and PyMem_Free

20121107
* updates msvscpp 2010 build

20121104
* pyvshadow: fixes for 32-bit build

20121103
* fix in spec file
* updated dependencies

20121102
* pyvshadow: fix for Mac OS X build
* updated msvscpp files

20121017
* code clean up

20121016
* pyvshadow: bug fixes
  - Missing Py_None increment reference
  - added increment/decrement reference of volume object in store

20121012
* pyvshadow: added creation time as integer function

20121011
* minor changes

20121010
* made get store more restrictive

20121009
* added store get size function for python binding

20121008
* updated dpkg and spec files

20121007
* added store get offset function
* worked on Python bindings
* fix for dpkg files docs

20121006
* worked on Python bindings

20120922
* small changes

20120921
* improved dealing with NTFS backup header for volume size detection

20120915
* improved catalog entry handling
* changes for libcdata
* worked on man pages
* worked on removing hard coded block size
* code clean up
* worked on support for catalog size corruption issue

20120914
* improved catalog entry handling
* added -X option to vshadowmount to pass options to fuse layer

20120912
* vshadowmount: replaced EBADFD by EIO for OSX

20120911
* fixed bug regarding missing block size correction

20120905
* updated libbfio

20120904
* code clean up

20120903
* changed overlay handling
* code clean up

20120902
* replaced data types by libcdata

20120901
* split off block list and block tree usage of block descriptors
  - block tree can override info in block descriptors while block list should not
* moved overlays from array to single block descriptor
* removed metadata offset list

20120831
* fixes and improvements in block tree
* changed initial bitmap to current bitmap
* code clean up

20120830
* fixes and improvements in btree

20120829
* fixes in btree
* added more debug info

20120828
* added more debug info

20120826
* code clean up
* initial version of msvscpp files

20120825
* improvement to handling not in-use blocks

20120824
* added support to handle duplicate reverse block tree blocks
* volume size is now determined from store descriptors in catalog
* code clean up

20120823
* worked on multi snapshot handling

20120822
* worked on block descriptors
* code clean up

20120821
* worked on block descriptors

20120820
* applied updates
* code clean

20120721
* added libcsplit
* applied updates
* switched library order in tools

20120603
* worked on block behavior

20120528
* worked on block behavior
* bug fix in mount handle

20120527
* worked on block behavior

20120526
* worked on block behavior

20120525
* worked on block behavior

20120524
* worked on block behavior
* re-introduced block wrap

20120523
* worked on block behavior

20120522
* worked on block behavior

20120520
* worked on block behavior

20120519
* worked on initial and previous bitmap

20120517
* worked on initial bitmap
* optimized reading bitmap
* fixed memory leaks
* worked on windows 8 support

20120515
* added support for volume without shadows

20120513
* fixed mingw build - linking of libbfio

20120511
* updated dependencies
* added fuse destroy function

20120415
* updated dependencies

vshadowtools/vshadowmount.c
/*
 * Mounts a Volume Service Snapshot (VSS) volume.
 *
 * Copyright (C) 2011-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"
#include "vshadowtools_getopt.h"
#include "vshadowtools_i18n.h"
#include "vshadowtools_libcerror.h"
#include "vshadowtools_libclocale.h"
#include "vshadowtools_libcnotify.h"
#include "vshadowtools_libvshadow.h"
#include "vshadowtools_output.h"
#include "vshadowtools_signal.h"
#include "vshadowtools_unused.h"

mount_handle_t *vshadowmount_mount_handle = NULL;
int vshadowmount_abort                    = 0;

/* Signal handler for vshadowmount
 */
void vshadowmount_signal_handler(
      vshadowtools_signal_t signal VSHADOWTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "vshadowmount_signal_handler";

	VSHADOWTOOLS_UNREFERENCED_PARAMETER( signal )

	vshadowmount_abort = 1;

	if( vshadowmount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     vshadowmount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use vshadowmount to mount a Volume Service Snapshot (VSS) volume.";

	vshadowtools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'o', "offset", "specify the volume offset in bytes" },
		{ 'v', NULL, "verbose output to stderr, while vshadowmount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "volume", "a Volume Service Snapshot (VSS) volume" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	libvshadow_error_t *error                    = NULL;
	system_character_t *option_offset            = NULL;
	const system_character_t *path_prefix        = NULL;
	system_character_t *source                   = NULL;
	char *program                                = "vshadowmount";
	system_integer_t option                      = 0;
	size_t path_prefix_size                      = 0;
	int number_of_options                        = (int) ( sizeof( options ) / sizeof( vshadowtools_option_t ) );
	int verbose                                  = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point              = NULL;
	int result                                   = 0;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations vshadowmount_fuse_operations;

	system_character_t *option_extended_options  = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                         = { program, NULL };
	struct fuse_args vshadowmount_fuse_arguments = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args vshadowmount_fuse_arguments = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *vshadowmount_fuse_channel  = NULL;
#endif
	struct fuse *vshadowmount_fuse_handle        = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS vshadowmount_dokan_operations;
	DOKAN_OPTIONS vshadowmount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "vshadowtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( vshadowtools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	vshadowtools_output_version_fprint(
	 stdout,
	 program );

	if( vshadowtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = vshadowtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				vshadowtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				vshadowtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'o':
				option_offset = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				vshadowtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source volume.\n" );

		vshadowtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind++ ];

	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		vshadowtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ optind ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libvshadow_notify_set_stream(
	 stderr,
	 NULL );
	libvshadow_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &vshadowmount_mount_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
	if( option_offset != NULL )
	{
		if( mount_handle_set_offset(
		     vshadowmount_mount_handle,
		     option_offset,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set volume offset.\n" );

			goto on_error;
		}
	}
#if defined( WINAPI )
	path_prefix = _SYSTEM_STRING( "\\VSS" );
#else
	path_prefix = _SYSTEM_STRING( "/vss" );
#endif
	path_prefix_size = 1 + system_string_length(
	                        path_prefix );

	if( mount_handle_set_path_prefix(
	     vshadowmount_mount_handle,
	     path_prefix,
	     path_prefix_size,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set path prefix.\n" );

		goto on_error;
	}
	if( mount_handle_open(
	     vshadowmount_mount_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source volume\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		vshadowmount_fuse_arguments.argc = 0;
		vshadowmount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &vshadowmount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &vshadowmount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &vshadowmount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &vshadowmount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	vshadowmount_fuse_operations.open       = &mount_fuse_open;
	vshadowmount_fuse_operations.read       = &mount_fuse_read;
	vshadowmount_fuse_operations.release    = &mount_fuse_release;
	vshadowmount_fuse_operations.opendir    = &mount_fuse_opendir;
	vshadowmount_fuse_operations.readdir    = &mount_fuse_readdir;
	vshadowmount_fuse_operations.releasedir = &mount_fuse_releasedir;
	vshadowmount_fuse_operations.getattr    = &mount_fuse_getattr;
	vshadowmount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	vshadowmount_fuse_handle = fuse_new(
	                            &vshadowmount_fuse_arguments,
	                            &vshadowmount_fuse_operations,
	                            sizeof( struct fuse_operations ),
	                            vshadowmount_mount_handle );

	if( vshadowmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          vshadowmount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	vshadowmount_fuse_channel = fuse_mount(
	                             mount_point,
	                             &vshadowmount_fuse_arguments );

	if( vshadowmount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	vshadowmount_fuse_handle = fuse_new(
	                            vshadowmount_fuse_channel,
	                            &vshadowmount_fuse_arguments,
	                            &vshadowmount_fuse_operations,
	                            sizeof( struct fuse_operations ),
	                            vshadowmount_mount_handle );

	if( vshadowmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          vshadowmount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
	fuse_unmount(
	 vshadowmount_fuse_handle );

	fuse_destroy(
	 vshadowmount_fuse_handle );

	fuse_opt_free_args(
	 &vshadowmount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &vshadowmount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &vshadowmount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	vshadowmount_dokan_options.Version    = DOKAN_VERSION;
	vshadowmount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	vshadowmount_dokan_options.SingleThread = TRUE;
#else
	vshadowmount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		vshadowmount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		vshadowmount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	vshadowmount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	vshadowmount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	vshadowmount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	vshadowmount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	vshadowmount_dokan_operations.CreateDirectory      = NULL;
	vshadowmount_dokan_operations.Cleanup              = NULL;
	vshadowmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	vshadowmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	vshadowmount_dokan_operations.WriteFile            = NULL;
	vshadowmount_dokan_operations.FlushFileBuffers     = NULL;
	vshadowmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	vshadowmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	vshadowmount_dokan_operations.FindFilesWithPattern = NULL;
	vshadowmount_dokan_operations.SetFileAttributes    = NULL;
	vshadowmount_dokan_operations.SetFileTime          = NULL;
	vshadowmount_dokan_operations.DeleteFile           = NULL;
	vshadowmount_dokan_operations.DeleteDirectory      = NULL;
	vshadowmount_dokan_operations.MoveFile             = NULL;
	vshadowmount_dokan_operations.SetEndOfFile         = NULL;
	vshadowmount_dokan_operations.SetAllocationSize    = NULL;
	vshadowmount_dokan_operations.LockFile             = NULL;
	vshadowmount_dokan_operations.UnlockFile           = NULL;
	vshadowmount_dokan_operations.GetFileSecurity      = NULL;
	vshadowmount_dokan_operations.SetFileSecurity      = NULL;
	vshadowmount_dokan_operations.GetDiskFreeSpace     = NULL;
	vshadowmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	vshadowmount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	vshadowmount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	vshadowmount_dokan_operations.Cleanup              = NULL;
	vshadowmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	vshadowmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	vshadowmount_dokan_operations.WriteFile            = NULL;
	vshadowmount_dokan_operations.FlushFileBuffers     = NULL;
	vshadowmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	vshadowmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	vshadowmount_dokan_operations.FindFilesWithPattern = NULL;
	vshadowmount_dokan_operations.SetFileAttributes    = NULL;
	vshadowmount_dokan_operations.SetFileTime          = NULL;
	vshadowmount_dokan_operations.DeleteFile           = NULL;
	vshadowmount_dokan_operations.DeleteDirectory      = NULL;
	vshadowmount_dokan_operations.MoveFile             = NULL;
	vshadowmount_dokan_operations.SetEndOfFile         = NULL;
	vshadowmount_dokan_operations.SetAllocationSize    = NULL;
	vshadowmount_dokan_operations.LockFile             = NULL;
	vshadowmount_dokan_operations.UnlockFile           = NULL;
	vshadowmount_dokan_operations.GetFileSecurity      = NULL;
	vshadowmount_dokan_operations.SetFileSecurity      = NULL;
	vshadowmount_dokan_operations.GetDiskFreeSpace     = NULL;
	vshadowmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	vshadowmount_dokan_operations.Unmounted            = NULL;
	vshadowmount_dokan_operations.FindStreams          = NULL;
	vshadowmount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &vshadowmount_dokan_options,
	          &vshadowmount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount Volume Shadow Snapshot (VSS) format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( vshadowmount_fuse_handle != NULL )
	{
		fuse_destroy(
		 vshadowmount_fuse_handle );
	}
	fuse_opt_free_args(
	 &vshadowmount_fuse_arguments );
#endif
	if( vshadowmount_mount_handle != NULL )
	{
		mount_handle_free(
		 &vshadowmount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

